src/bitmap/backend/atomic_bitmap_mmap.rs

```diff
@@ -0,0 +1,608 @@
+        // However, since each memory region requires a bitmap relative to them, we have to
+        // adjust the offset and size, in number of pages, of this region.
+        let offset_pages = page_number(region_start_addr);
+        let size_page = page_number(region_len);
+
+        Ok(Self {
+            logmem,
+            pages_before_region: offset_pages,
+            number_of_pages: size_page,
+        })
+    }
+}
+
+// Just a handy object to manage the RAII of the mmap region.
+#[derive(Debug)]
+struct MmapLogReg {
+    addr: *const AtomicU8,
+    len: usize,
+}
+
+// SAFETY: Send is not automatically implemented because the raw pointer.
+// No one besides `MmapLogReg` has the raw pointer, so we can safely transfer it to another thread.
+unsafe impl Send for MmapLogReg {}
+
+// SAFETY: Sync is not automatically implemented because the raw pointer.
+// `MmapLogReg` doesn't have any interior mutability and all access to `&AtomicU8`
+// are done through atomic operations.
+unsafe impl Sync for MmapLogReg {}
+
+impl MmapLogReg {
+    fn from_file(fd: BorrowedFd, offset: isize, len: usize) -> io::Result<Self> {
+        #[cfg(not(miri))]
+        // SAFETY: `fd` is a valid file descriptor and we are not using `libc::MAP_FIXED`.
+        let addr = unsafe {
+            libc::mmap(
+                ptr::null_mut(),
+                len as libc::size_t,
+                libc::PROT_READ | libc::PROT_WRITE,
+                libc::MAP_SHARED,
+                fd.as_raw_fd(),
+                offset as libc::off_t,
+            )
+        };
+
+        #[cfg(not(miri))]
+        if addr == libc::MAP_FAILED {
+            return Err(io::Error::last_os_error());
+        }
+
+        #[cfg(miri)]
+        if len == 0 {
+            return Err(io::Error::from_raw_os_error(libc::EINVAL));
+        }
+
+        // Miri does not support the mmap syscall, so we use rust's allocator for miri tests
+        #[cfg(miri)]
+        let addr = unsafe {
+            std::alloc::alloc_zeroed(std::alloc::Layout::from_size_align(len, 1).unwrap())
+        };
+
+        Ok(Self {
+            addr: addr as *const AtomicU8,
+            len,
+        })
+    }
+
+    fn base_ptr(&self) -> *const AtomicU8 {
+        self.addr
+    }
+}
+
+impl Index<usize> for MmapLogReg {
+    type Output = AtomicU8;
+
+    // It's ok to get a reference to an atomic value.
+    fn index(&self, index: usize) -> &Self::Output {
+        assert!(index < self.len);
+        // Note: Instaed of `&*` we can use `AtomicU8::from_ptr()` as soon it gets stabilized.
+        // SAFETY: `self.addr` + `index` doesn't wrap around and is contained within the mapped
+        // memory region.
+        unsafe { &*self.addr.add(index) }
+    }
+}
+
+impl Drop for MmapLogReg {
+    fn drop(&mut self) {
+        // SAFETY: `addr` is non-null and properly aligned, also we are sure that this
+        // is the last reference alive and/or we have an exclusive access to this object.
```

Comment:
First, there are different safety concerns for `munmap()` and `dealloc()`, I believe. That’s another reason why this implementation should be split more explicitly.

Second, as for `munmap`, I don’t think passing NULL or unaligned values would hurt the safety (`munmap()` would just fail), I think what’s important is that we got `self.addr` and `self.len` relate to the `mmap()` call in `Self::from_file()`.